### ISSUE TITLE
Feature Suggestions

### ISSUE THREAD

**Issue by ideosyncrasov** [borodust/trivial-gamekit#9]
Hey borodust, as promised here are some suggestions for features which I would like to see in trivial-gamekit (explanations below):

1. Add the ability to play looped sound.
2. Add `with-new-origin`.
3. Add the ability to draw to buffers (and be able to combine them).

*Regarding 1:* This would be e.g. useful for in-game music. It seems to me that this functionality is already present in `cl-bodge` and that doing 

    (let ((sound (gamekit::resource-by-id :sound-id)))
      (setf (gamekit::audio-looped-p sound) t)
      (gamekit:play :sound-id))

should do the trick, but that spawns an error.

*Regarding 2:* It would be useful if one could change the origin of the coordinate system. One convenient way would be a macro `with-new-origin` such that

    (with-new-origin (new-origin)
      (gamekit:draw-rect origin w h :fill-paint *black*))

would have the same effect as

    (gamekit:draw-rect (gamekit:add origin new-origin) w h :fill-paint *black*)

This seems useful for code decoupling.

*Regarding 3:* If you draw "directly" to the screen, you must be careful to draw things in order. Coupling your code to the drawing order is easy for very simple graphics, but once you have several visibility layers and have effects on top of that, it gets complicated. It would be therefore useful to think if one could draw to off-screen buffers, and would also be able to combine buffers arithmetically (e.g. masks, overlays etc.)

**Comment by borodust** [comment]
Indeed, 1) is supported by `cl-bodge`. I'll introduce looped playback into `trivial-gamekit` and also function for stopping it.

2) is present in `cl-bodge` too. It is (and all canvas-related methods) can be found in `ge.vg` package. For transferring origin you would use #'translate-canvas. This is a fairly simple concept, so yes, I'll reimport transform functions (translate, rotate, shear, scale) into `trivial-gamekit`, possibly renaming them for simplicity.

3) is an advanced technique. Unfortunately, I'm not planning to add any specific interfaces for it into `trivial-gamekit`. Hopefully, FBO and off-screen rendering is actually already supported by `cl-bodge`, so one could use its capabilities for that if only it had any documentation :) If heavily needed (personally, i have no problem with the idea of organizing virtual drawing layers for ordered rendering), I can explain how to use `cl-bodge`'s FBOs to render into texture.

**Comment by ideosyncrasov** [comment]
**Regarding (1):** Is this supposed to work already? If I evaluate `(gamekit::audio-looped-p (gamekit::resource-by-id :sound-id))` it gives me the following error:

```
:LOOPED fell through ECASE expression.
Wanted one of (:GAIN :PITCH :MIN-GAIN :MAX-GAIN
               :REFERENCE-DISTANCE :SEC-OFFSET :ROLLOFF-FACTOR
               :MAX-DISTANCE :CONE-INNER-ANGLE :CONE-OUTER-ANGLE
               :CONE-OUTER-GAIN :SAMPLE-OFFSET :BYTE-OFFSET
               :LOOPING :SOURCE-RELATIVE :SOURCE-TYPE :BUFFER
               :BUFFERS-QUEUED :BUFFERS-PROCESSED :SOURCE-STATE
               :POSITION :VELOCITY :DIRECTION).
   [Condition of type SB-KERNEL:CASE-FAILURE]
```

**Regarding (2):** Having the ability to transform the canvas is useful, but it would also be convenient to have a macro that allows one to limit the effects of the transformations to a given code body in a fool-proof way. It seems to me, the macro `with-pushed-canvas` does exactly that (pushing a copy of the canvas on a stack and popping it at the end I suppose), so maybe that macro could also be exposed.

**Regarding (3):** I'm aware that this is an advanced feature, that's why I was hesistant to even include it in this list. But it is a feature that I would have found useful, so that's why I included it anyway. Now, if every feature of `cl-bodge` would be exposed directly by `trivial-gamekit`, then the question would arise why not use `cl-bodge` directly. So feature bloat should definitely be avoided.

But I think it might be a compromise to *amend the documentation* by a section like "How to use advanced features from `cl-bodge`" that

- makes people aware that `trivial-gamekit` is really just a simplified interface to `cl-bodge`, and that the features of the latter are basically all still available from within the former
- explains with an example (doesn't have to be FBOs) how one can extend `trivial-gamekit` by features from `cl-bodge` as needed

For example, when I noticed I had to draw parts of images and that feature wasn't readily available, I was afraid I would need to switch engines. But after I became aware that the whole of `cl-bodge` is accessible from within gamekit simply by using double-colons instead of colons, coding in the feature I needed was trivial.
